Main

Emmankoko opened the pull request.

Emmankoko committed:
recent pushes

Emmankoko committed:
recent pushes

Geod24 commented on source/stdcpp/vector.d:
```diff
@@ -750,10 +768,84 @@ extern(D):
+	//hard coding fill_n for current testing, will organize into functions
+	//for optimal organization after test passes well.
+	
+	if(__traits(isArithmetic, T))
```
This might need to be `static if`. Since the condition is always known at compiler time.

Geod24 commented on source/stdcpp/vector.d:
```diff
@@ -750,10 +768,84 @@ extern(D):
+	//finish pointer moves to end of storage after initialization
+	
+	this._M_finish = this._M_end_of_storage; 
+	}
```
Please configure your editor to automatically indent by 4 spaces.

Geod24 commented on source/stdcpp/vector.d:
```diff
@@ -70,15 +70,28 @@ extern(D):
+commenting opOpAssign, append and opAssign out which uses insert 
+which is currently facing a mangling issue for gcc and focus to get the basics
+done and working 
+*********************************************************************/
```
Can you show me the mangling issue ?

Geod24 commented on source/stdcpp/vector.d:
```diff
@@ -750,10 +768,84 @@ extern(D):
+
+	// vector(n) constructor
+
+	this()(size_t __n)
```
Note that this constructor is templated, something that is not possible in C++. So this does not mangle to the constructor as it should.
Try compiling the following:
```C++
#include <vector>

int main ()
{
    std::vector<int> vec;
    return 0;
}
```

And then `nm a.out`, you'll get the following (removed irrelevant output):
```
0000000000001481 W _ZNSt12_Destroy_auxILb1EE9__destroyIPiEEvT_S3_
000000000000121e W _ZNSt12_Vector_baseIiSaIiEE12_Vector_implC1Ev
000000000000121e W _ZNSt12_Vector_baseIiSaIiEE12_Vector_implC2Ev
00000000000011be W _ZNSt12_Vector_baseIiSaIiEE12_Vector_implD1Ev
00000000000011be W _ZNSt12_Vector_baseIiSaIiEE12_Vector_implD2Ev
0000000000001366 W _ZNSt12_Vector_baseIiSaIiEE13_M_deallocateEPim
0000000000001324 W _ZNSt12_Vector_baseIiSaIiEE17_Vector_impl_dataC1Ev
0000000000001324 W _ZNSt12_Vector_baseIiSaIiEE17_Vector_impl_dataC2Ev
00000000000013a0 W _ZNSt12_Vector_baseIiSaIiEE19_M_get_Tp_allocatorEv
00000000000011de W _ZNSt12_Vector_baseIiSaIiEEC1Ev
00000000000011de W _ZNSt12_Vector_baseIiSaIiEEC2Ev
000000000000126a W _ZNSt12_Vector_baseIiSaIiEED1Ev
000000000000126a W _ZNSt12_Vector_baseIiSaIiEED2Ev
00000000000013ef W _ZNSt16allocator_traitsISaIiEE10deallocateERS0_Pim
00000000000011fe W _ZNSt6vectorIiSaIiEEC1Ev
00000000000011fe W _ZNSt6vectorIiSaIiEEC2Ev
00000000000012bc W _ZNSt6vectorIiSaIiEED1Ev
00000000000012bc W _ZNSt6vectorIiSaIiEED2Ev
0000000000001421 W _ZSt8_DestroyIPiEvT_S1_
00000000000013b2 W _ZSt8_DestroyIPiiEvT_S1_RSaIT0_E
```

If you run c++filt on this:
```
0000000000001481 W void std::_Destroy_aux<true>::__destroy<int*>(int*, int*)
000000000000121e W std::_Vector_base<int, std::allocator<int> >::_Vector_impl::_Vector_impl()
000000000000121e W std::_Vector_base<int, std::allocator<int> >::_Vector_impl::_Vector_impl()
00000000000011be W std::_Vector_base<int, std::allocator<int> >::_Vector_impl::~_Vector_impl()
00000000000011be W std::_Vector_base<int, std::allocator<int> >::_Vector_impl::~_Vector_impl()
0000000000001366 W std::_Vector_base<int, std::allocator<int> >::_M_deallocate(int*, unsigned long)
0000000000001324 W std::_Vector_base<int, std::allocator<int> >::_Vector_impl_data::_Vector_impl_data()
0000000000001324 W std::_Vector_base<int, std::allocator<int> >::_Vector_impl_data::_Vector_impl_data()
00000000000013a0 W std::_Vector_base<int, std::allocator<int> >::_M_get_Tp_allocator()
00000000000011de W std::_Vector_base<int, std::allocator<int> >::_Vector_base()
00000000000011de W std::_Vector_base<int, std::allocator<int> >::_Vector_base()
000000000000126a W std::_Vector_base<int, std::allocator<int> >::~_Vector_base()
000000000000126a W std::_Vector_base<int, std::allocator<int> >::~_Vector_base()
00000000000013ef W std::allocator_traits<std::allocator<int> >::deallocate(std::allocator<int>&, int*, unsigned long)
00000000000011fe W std::vector<int, std::allocator<int> >::vector()
00000000000011fe W std::vector<int, std::allocator<int> >::vector()
00000000000012bc W std::vector<int, std::allocator<int> >::~vector()
00000000000012bc W std::vector<int, std::allocator<int> >::~vector()
0000000000001421 W void std::_Destroy<int*>(int*, int*)
00000000000013b2 W void std::_Destroy<int*, int>(int*, int*, std::allocator<int>&)
```

So as you can see , the C++ template emits the ctor, you just have to bind it.

Geod24 reviewed (changes requested):
Made a few comments. Also make sure to include a proper title / commit message in the future.

Emmankoko committed:
check for vector(n) instantiation and push_back bindings

Emmankoko committed:
Undo "check for vector(n) instantiation and push_back bindings"

This reverts commit d9df6ad9f71d6ca1b285b5339972a33dbce77737.

Emmankoko committed:
undo "recent pushes"

This reverts commit 41c95dccf95fe6286ae9b943af7384680ab4caf7.

Emmankoko committed:
undo "recent pushes"

This reverts commit ae0fd795f02ab3ae7fadc38a4ba6c42dd43c255c.

Emmankoko committed:
move push_back to microsoft runtime

Emmankoko committed:
move opAssign to microsoft runtime

Emmankoko committed:
move opOpAssign and append to microsoft runtime

Emmankoko committed:
size definition for gcc runtime

Emmankoko committed:
as_array function defintion for gcc runtime

Emmankoko committed:
push_back bindings for lvalues

Emmankoko committed:
vector(n) constructor

Emmankoko committed:
simple unittest for vector on gcc

Geod24 committed:
Add vector.cpp and link to it in unittest

We need an explicit instantiation of vector to link against,
so that C++ symbols are available. It's an extra step,
but any sizeable project will have this requirement one
way or another, and it is the most foolproof approach
(the other approach is to reproduce what std::vector
does in its C++ code while managing the 3 pointers).

For the moment, just add an extras/ folder and assume
unittests will need to add an explicit instantiation
when using symbols. In the future we can make that code
more generic, or even better, auto-generated.

Geod24 commented:
@Emmankoko : This is what you are missing:
https://github.com/dlang/stdcpp/commit/9c780ac69a495efdbe22ebf802d1112f7af9e794
Do a `git fetch` and then `git cherry-pick 9c780ac` on your branch to import it.

Emmankoko committed:
bindings for at and capacity for gcc runtime

Emmankoko committed:
vector(n, value) constructor for gcc

Emmankoko committed:
reserve for gcc

Emmankoko committed:
resize for gcc

Emmankoko committed:
empty() for gcc

Emmankoko committed:
clear() for gcc

Emmankoko committed:
pop_back() for gcc runtime

Emmankoko committed:
unittest for newly added member functions

Geod24 commented on source/stdcpp/test/vector.d:
```diff
@@ -5,13 +5,38 @@
+import stdcpp.allocator;
+int a = 42;
+int b = 9;
+allocator!int alloc_instance = allocator!(int).init;
```
We need to remove this. I think a D constructor that forwards to the C++ constructor makes sense.

Geod24 commented on source/stdcpp/test/vector.d:
```diff
@@ -5,13 +5,38 @@
-
 import stdcpp.vector;
+import stdcpp.allocator;
+int a = 42;
+int b = 9;
```
Remove or move to unittest scope

Emmankoko replied:
removed

Geod24 commented on source/stdcpp/test/vector.d:
```diff
@@ -5,13 +5,38 @@
+unittest
+{
+    auto vec = vector!int(4, alloc_instance);
+    vec.push_back(a);
```
We need to accept rvalue

Emmankoko replied:
got it working now

Emmankoko committed:
assign function for gcc

Geod24 commented on source/stdcpp/vector.d:
```diff
@@ -72,13 +72,9 @@ extern(D):
-    {
-        clear();
-        reserve(array.length);
-        insert(0, array);
```
```D
version (CppRuntime_Microsoft)
    insert(0, array);
else
    // Do something else
```

Emmankoko committed:
push_back and assign function for rvalues passing with their test cases for rvalues

Emmankoko committed:
copy and allocator constructor,defaulted allocator parameter and simple unittest for passing rvalues to constructors

Emmankoko committed:
simple fixes on std::string(we can use std_string now for instantiations)

Emmankoko committed:
additions(iterator functions and conditional compilation settled) and their tests

Geod24 commented on source/stdcpp/string.d:
```diff
@@ -108,10 +108,10 @@ extern(C++, (StdNamespace)) struct char_traits(CharT)
 
-// I don't think we can have these here, otherwise symbols are emit to druntime, and we don't want that...
-//alias std_string = basic_string!char;
+// I think we can have these here now, since the project have been moved from druntime
```
Just replace the comment with `///` (so that it ends up in documentation)

Emmankoko committed:
additional test iterations

Emmankoko committed:
early stage binding for std::list, targetting GNU gcc pre-c++11 and post-c++11 standards.
tests added for bindings and more to come.

Emmankoko committed:
configuration correction

Emmankoko committed:
dub configuration correction

Emmankoko committed:
fixing improper spacing and alignment

Emmankoko committed:
fixing improper spacing and alignment

Emmankoko committed:
setting up macOS and clang testing. dtor in list commented out for now as it breaks macOS(clang) testing for vector

Emmankoko committed:
undo "setting up macOS and clang testing. dtor in list commented out for now as it breaks macOS(clang) testing for vector"

This reverts commit 35f5ba30434e39787e02485f97e6c808d91192a6.

Emmankoko committed:
using clang to produce symbols for tests on macOS and ubuntu

Emmankoko committed:
test for runtime clang on ubuntu

Emmankoko committed:
test for runtime clang on ubuntu

Emmankoko committed:
clang 9.0.0 tests
since clang 13.0.0 is not setting up on macos-11 runner

Emmankoko committed:
macos clang 8 setup test

Emmankoko committed:
macos clang 8 setup test(diff arch)

Emmankoko committed:
macos clang setup test

Emmankoko committed:
macos clang setup test

Emmankoko committed:
macos-11 clang setup test

Emmankoko committed:
clang setup on macos-11 test

Emmankoko committed:
clang setup on macos-11 test

Emmankoko committed:
clang setup on macos-11 test

Emmankoko committed:
isolated list.d and string.d to avoid hindrance for macos clang testing

Emmankoko committed:
test for clang-13

Emmankoko committed:
test for OSX

Emmankoko committed:
test for macos

Emmankoko committed:
test for macos

Emmankoko committed:
test for macos

Emmankoko committed:
test for macos

Emmankoko committed:
placing things back

Emmankoko committed:
placing things back

Emmankoko committed:
placing things back

Emmankoko committed:
putting things back

Geod24 commented on .github/workflows/e2e.yml:
```diff
@@ -23,8 +23,8 @@ on:
   push:
     branches:
-      - v*.*.x
-      # Use this branch name in your fork to test changes
+      - main
+       #Use this branch name in your fork to test changes
```
Debugging artifacts

Geod24 commented on dub.json:
```diff
@@ -13,6 +13,12 @@
         },
         {
             "name": "unittest",
+            "lflags": [ "-lstdc++" ],
+            "preGenerateCommands": [
+                "clang++ -c extras/vector.cpp -o extras/vector.o",
+				"clang++ -c extras/list.cpp -o extras/list.o"
+            ],
+            "sourceFiles": [ "extras/*.o" ]
```
Only works for clang, we need a way to test both clang++ and g++.

Ideas:
- A configuration that is `unittest-clang` and one `unittest-gcc` but then try to have a sane default.

Geod24 commented on source/stdcpp/list.d:
```diff
@@ -0,0 +1,318 @@
+
+import stdcpp.allocator;
+
+version(CppRuntime_Gcc)
```
https://dlang.org/dstyle.html

Geod24 commented on source/stdcpp/list.d:
```diff
@@ -0,0 +1,318 @@
+
+	alias listNamespace = StdNamespace;
+}
+enum def {value}; //for later use
```
Can't call C++ default constructor from D when doing this:
```
list!int my_list;
```

Options:
`@disable this() @safe pure nothrow @nogc scope;`

But also defining a constructor hides the default constructor.

Can do:

`static list make ();`

Geod24 commented on source/stdcpp/list.d:
```diff
@@ -0,0 +1,318 @@
+
+
+
+	@disable this();
```
Attributes

Geod24 commented on source/stdcpp/list.d:
```diff
@@ -0,0 +1,318 @@
+				this(n, type_instance, alloc_instance);
+			}
+
+//			~this();
```
Clean any artifacts

Geod24 commented:
Did a few adjustments and moved to #6
